crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -437,46 +609,21 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 
     fn infer_name_expression(&mut self, name: &ast::ExprName) -> Type<'db> {
-        let ast::ExprName { range: _, id, ctx } = name;
+        let ast::ExprName {
+            range: _,
+            id: _,
+            ctx,
+        } = name;
 
         match ctx {
             ExprContext::Load => {
-                let ancestors = self.index.ancestor_scopes(self.file_scope_id);
-
-                for (ancestor_id, _) in ancestors {
-                    // TODO: Skip over class scopes unless the they are a immediately-nested type param scope.
-                    // TODO: Support built-ins
-
-                    let (symbol_table, ancestor_scope) = if ancestor_id == self.file_scope_id {
-                        (Cow::Borrowed(&self.symbol_table), None)
-                    } else {
-                        let ancestor_scope = ancestor_id.to_scope_id(self.db, self.file_id);
-                        (
-                            Cow::Owned(symbol_table(self.db, ancestor_scope)),
-                            Some(ancestor_scope),
-                        )
-                    };
-
-                    if let Some(symbol_id) = symbol_table.symbol_id_by_name(id) {
-                        let symbol = symbol_table.symbol(symbol_id);
-
-                        if !symbol.is_defined() {
-                            continue;
-                        }
-
-                        return if let Some(ancestor_scope) = ancestor_scope {
-                            let types = infer_types(self.db, ancestor_scope);
-                            types.symbol_ty(symbol_id)
-                        } else {
-                            self.local_definition_ty(symbol_id)
-                        };
-                    }
-                }
-                Type::Unknown
+                let use_def = use_def_map(self.db, self.scope);
```

Comment:
We should use the `index` directly here because using the `use_def_map` for better "isolation" doesn't give us anything because this query already depends on the AST and the semantic index.

Using the query here actually results in a negative performance impact because we now pay the cost of another concurrent hash map lookup